proposals/NNNN-member-import-visibility.md

```diff
@@ -0,0 +1,153 @@
+# Member import visibility
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Allan Shortlidge](https://github.com/tshortli)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Bug: [apple/swift#46493](https://github.com/apple/swift/issues/46493)
+* Implementation: [apple/swift#72974](https://github.com/apple/swift/pull/72974), [apple/swift#73063](https://github.com/apple/swift/pull/73063)
+* Upcoming Feature Flag: `MemberImportVisibility`
+* Review: TBD
+
+## Introduction
+
+In Swift, there are rules dictating whether the name of a declaration in another module is considered in scope. For example, if you have a program that uses the `swift-algorithms` package and you want to use the global function [chain()](https://github.com/apple/swift-algorithms/blob/33abb694280321a84aa7dc9806de284afb8ca226/Sources/Algorithms/Chain.swift#L287) then you must write `import Algorithms` in the file that references that function or the compiler will consider it out of scope:
+
+``` swift
+// Missing 'import Algorithms'
+let chained = chain([1], [2]) // error: Cannot find 'chain' in scope
+```
+
+The visibility rules for a member declaration, such as a method declared inside of a struct, are different though. When resolving a name to a member declaration, the member is in scope even if the module introducing the member is only *transitively* imported. A transitively imported module could be imported directly in another source file, or it could be a dependency of some direct dependency of your program. This inconsistency may be best understood as a subtle bug rather than an intentional design decision, and in a lot of Swift code it goes unnoticed. However, the import rules for members become more surprising when you consider the members of extensions, since an extension and its nominal type can be declared in different modules. 
+
+This proposal unifies the behavior of name lookup by changing the rules to bring both top-level declarations and members into scope using the same criteria.
+
+## Motivation
+
+Suppose you have a program depends on a library named `RecipeKit`. The library interface looks like this:
+
+```swift
+// RecipeKit interface
+
+public struct Recipe { /*...*/ }
+
+extension String {
+  /// Returns the recipe represented by this string.
+  public func parse() -> Recipe?
+}
+```
+
+To start, your program contains a single source file `main.swift` that imports `RecipeKit`:
+
+```swift
+// main.swift
+import RecipeKit
+
+let recipe = "2 slices of bread, 1.5 tbs peanut butter".parse()
+```
+
+Later, you decide to integrate with a new library named `GroceryKit` which happens to also declares its own `parse()` method in an extension on `String`:
+
+```swift
+// GroceryKit interface
+public struct GroceryList { /*...*/ }
+
+extension String {
+  /// Returns the grocery list represented by this string.
+  public func parse() -> GroceryList?
+}
+
+```
+
+You add a second file that imports `GroceryKit`:
```

Comment:
Nit: It took me a few reads to identify _to which module_ this second file is being added. By the phrasing of the overall problem statement it must be to StoreKit, but a little clarification of the exact structure built up in this hypothetical could help here.

Reply:
I'm a bit confused, there is no `StoreKit`. I haven't named the module that the source files belong to, it's just an anonymous program that you are building which has `RecipeKit` and `GroceryKit` as external dependencies.

Reply:
I'll try to tweak this language to clarify that the second file is being added to your program. Maybe calling it "your app" would help?